File: lib/http.js
import Gio from 'gi://Gio';
import GLib from 'gi://GLib';
import Soup from 'gi://Soup';

Gio._promisify(Soup.Session.prototype, 'send_and_read_async', 'send_and_read_finish');

export class HttpError extends Error {
    constructor(message, {statusCode = 0, body = null} = {}) {
        super(String(message));
        this.name = 'HttpError';
        this.statusCode = statusCode;
        this.body = body;
    }

    get isAuthError() {
        return this.statusCode === 401 || this.statusCode === 403;
    }
}

export class HttpClient {
    constructor({timeout = 30} = {}) {
        this._session = new Soup.Session({timeout});
    }

    destroy() {
        this._session.abort();
    }

    async getJson(url, {headers = {}} = {}) {
        return this._requestJson('GET', url, {headers});
    }

    async postJson(url, body, {headers = {}} = {}) {
        return this._requestJson('POST', url, {
            headers: {
                'Content-Type': 'application/json',
                ...headers,
            },
            body: typeof body === 'string' ? body : JSON.stringify(body ?? {}),
        });
    }

    async postForm(url, fields, {headers = {}} = {}) {
        const form = Object.entries(fields)
            .map(([key, value]) =>
                `${encodeURIComponent(key)}=${encodeURIComponent(String(value ?? ''))}`)
            .join('&');

        return this._requestJson('POST', url, {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                ...headers,
            },
            body: form,
        });
    }

    async _requestJson(method, url, {headers = {}, body = null} = {}) {
        const message = Soup.Message.new(method, url);
        const requestHeaders = message.get_request_headers();
        for (const [key, value] of Object.entries(headers)) {
            if (value !== undefined && value !== null && value !== '')
                requestHeaders.append(key, String(value));
        }

        if (body !== null) {
            const bytes = new TextEncoder().encode(body);
            message.set_request_body_from_bytes(
                headers['Content-Type'] ?? 'application/octet-stream',
                GLib.Bytes.new(bytes),
            );
        }

        const responseBytes = await this._session.send_and_read_async(
            message,
            GLib.PRIORITY_DEFAULT,
            null,
        );

        const statusCode = message.get_status();
        const text = decodeBytes(responseBytes);
        let payload = null;
        if (text) {
            try {
                payload = JSON.parse(text);
            } catch (error) {
                if (statusCode >= 200 && statusCode < 300) {
                    throw new HttpError(`Invalid JSON from ${url}`, {
                        statusCode,
                        body: text,
                    });
                }
            }
        }

        if (statusCode < 200 || statusCode >= 300) {
            const messageText = extractErrorMessage(payload) ||
                `Request failed with HTTP ${statusCode}.`;
            throw new HttpError(messageText, {statusCode, body: payload ?? text});
        }

        return payload;
    }
}

export function decodeBytes(bytes) {
    const data = bytes?.toArray?.() ?? bytes?.get_data?.() ?? [];
    return new TextDecoder().decode(data);
}

function extractErrorMessage(payload) {
    if (!payload)
        return '';
    if (typeof payload === 'string')
        return payload.trim();
    if (typeof payload !== 'object')
        return '';

    for (const value of [payload.message, payload.error, payload.detail, payload.title, payload.description]) {
        if (typeof value === 'string' && value.trim())
            return value.trim();
        if (value && typeof value === 'object') {
            if (typeof value.message === 'string' && value.message.trim())
                return value.message.trim();
        }
    }
    return '';
}

File: src/http.rs
//! Minimal blocking HTTP helpers (reqwest).

use anyhow::{Context, Result, anyhow};
use reqwest::blocking::{Client, Response};
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use serde::de::DeserializeOwned;
use serde_json::Value;
use std::collections::HashMap;
use std::time::Duration;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum HttpError {
    #[error("{message}")]
    Status {
        status: u16,
        message: String,
        body: Option<Value>,
    },
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl HttpError {
    pub fn is_auth_error(&self) -> bool {
        matches!(self, Self::Status { status: 401 | 403, .. })
    }
}

#[derive(Clone)]
pub struct HttpClient {
    inner: Client,
}

impl HttpClient {
    pub fn new() -> Result<Self> {
        let inner = Client::builder()
            .timeout(Duration::from_secs(30))
            .user_agent("usagenometer/0.1")
            .build()
            .context("failed to build HTTP client")?;
        Ok(Self { inner })
    }

    pub fn get_json<T: DeserializeOwned>(
        &self,
        url: &str,
        headers: &[(&str, &str)],
    ) -> Result<T, HttpError> {
        let response = self
            .inner
            .get(url)
            .headers(build_headers(headers)?)
            .send()
            .map_err(|e| HttpError::Other(e.into()))?;
        parse_json(response)
    }

    pub fn post_json<T: DeserializeOwned>(
        &self,
        url: &str,
        body: &Value,
        headers: &[(&str, &str)],
    ) -> Result<T, HttpError> {
        let response = self
            .inner
            .post(url)
            .headers(build_headers(headers)?)
            .json(body)
            .send()
            .map_err(|e| HttpError::Other(e.into()))?;
        parse_json(response)
    }

    pub fn post_form<T: DeserializeOwned>(
        &self,
        url: &str,
        form: &HashMap<&str, &str>,
    ) -> Result<T, HttpError> {
        let response = self
            .inner
            .post(url)
            .form(form)
            .send()
            .map_err(|e| HttpError::Other(e.into()))?;
        parse_json(response)
    }
}

fn build_headers(headers: &[(&str, &str)]) -> Result<HeaderMap, HttpError> {
    let mut map = HeaderMap::new();
    for (k, v) in headers {
        let name = HeaderName::from_bytes(k.as_bytes())
            .map_err(|e| HttpError::Other(anyhow!("bad header name {k}: {e}")))?;
        let value = HeaderValue::from_str(v)
            .map_err(|e| HttpError::Other(anyhow!("bad header value for {k}: {e}")))?;
        map.append(name, value);
    }
    Ok(map)
}

fn parse_json<T: DeserializeOwned>(response: Response) -> Result<T, HttpError> {
    let status = response.status().as_u16();
    let text = response
        .text()
        .map_err(|e| HttpError::Other(e.into()))?;
    let payload: Option<Value> = if text.trim().is_empty() {
        None
    } else {
        Some(serde_json::from_str(&text).map_err(|e| {
            HttpError::Other(anyhow!("invalid JSON: {e}"))
        })?)
    };

    if !(200..300).contains(&status) {
        let message = payload
            .as_ref()
            .and_then(extract_error_message)
            .unwrap_or_else(|| format!("Request failed with HTTP {status}."));
        return Err(HttpError::Status {
            status,
            message,
            body: payload,
        });
    }

    let value = payload.unwrap_or(Value::Null);
    serde_json::from_value(value).map_err(|e| HttpError::Other(anyhow!("JSON shape: {e}")))
}

fn extract_error_message(payload: &Value) -> Option<String> {
    for key in ["message", "error", "detail", "title"] {
        if let Some(s) = payload.get(key).and_then(|v| {
            if let Some(s) = v.as_str() {
                let t = s.trim();
                if !t.is_empty() {
                    return Some(t.to_string());
                }
            }
            if let Some(obj) = v.as_object() {
                for nested in ["message", "detail", "title", "code", "type"] {
                    if let Some(s) = obj.get(nested).and_then(|x| x.as_str()) {
                        let t = s.trim();
                        if !t.is_empty() {
                            return Some(t.to_string());
                        }
                    }
                }
            }
            None
        }) {
            return Some(s);
        }
    }
    None
}
